<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to `adaptMT` package • adaptMT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to `adaptMT` package">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">adaptMT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/adapt_demo.html">Introduction to `adaptMT` package</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lihualei71/adaptMT">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="adapt_demo_files/header-attrs-2.2/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Introduction to <code>adaptMT</code> package</h1>
                        <h4 class="author">Lihua Lei</h4>
            
            <h4 class="date">2020-07-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/lihualei71/adaptMT/blob/master/vignettes/adapt_demo.Rmd"><code>vignettes/adapt_demo.Rmd</code></a></small>
      <div class="hidden name"><code>adapt_demo.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><code>adaptMT</code> package is an <code>R</code> implementation of Adaptive P-value Thresholding <span class="citation">(Lei and Fithian 2016)</span>. <code>adaptMT</code> package has two main components: a generic interface that allows users to specify the working model and algorithms to fit them, as well as a pool of easy-to-use end-to-end wrappers. The former is captured by function <code>adapt</code>. The latter includes <code>adapt_glm</code>, <code>adapt_gam</code> and <code>adapt_glmnet</code> for using GLM, GAM and L1-regularized GLM.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># install the "adaptMT" package from github.</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="co"># will be submitted to CRAN very soon.</span></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="co"># devtools::install_github("lihualei71/adaptMT")</span></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"adaptMT"</span>)</span></code></pre></div>
<p>AdaPT is a flexible iterative protocol that allows almost arbitrary exploration of data while still guaranteeing the False Discovery Rate (FDR) control in finite samples under standard assumptions in literature. The pipeline of AdaPT is as follows:</p>
<embed src="flowchart_AdaPT.pdf" width="75%" height="250px" style="display: block; margin: auto;" type="application/pdf"></embed><p>The three red components allow unlimited user-flexibility, as the theoretical results in <span class="citation">(Lei and Fithian 2016)</span> guarantees the FDR control regardless of the implementation. To get rid of over-complication and minimize the user inputs, <code>adaptMT</code> package implements a slightly restrictive , though still flexible enough, pipeline which</p>
<ul>
<li><p>exploits a conditional two-groups model as the working model (for EM pipeline), i.e. <span class="math display">\[H_i \mid x_i \sim \pi(x_i), \quad  p_i \mid H_i, x_i\sim \left\{\begin{array}{ll} \mathrm{Unif}([0, 1]) &amp; H_i = 0\\ h(p_i; \mu(x_i)) &amp; H_i = 1\end{array}\right..\]</span> See Section 4.1 of <span class="citation">(Lei and Fithian 2016)</span> for details;</p></li>
<li><p>performs model selection via information criterion (BIC, AIC, etc.) based on the pseudo-likelihood from the EM algorithm; See Section 4.2 of <span class="citation">(Lei and Fithian 2016)</span> for details;</p></li>
<li><p>permits arbitrary algorithm to fit <span class="math inline">\(\pi(x)\)</span> and <span class="math inline">\(\mu(x)\)</span>.</p></li>
</ul>
<p>We refer the readers to <span class="citation">(Lei and Fithian 2016)</span> for more details.</p>
</div>
<div id="convenient-wrappers-in-adaptmt-package" class="section level1">
<h1 class="hasAnchor">
<a href="#convenient-wrappers-in-adaptmt-package" class="anchor"></a>Convenient Wrappers in <code>adaptMT</code> Package</h1>
<p><code>adaptMT</code> package provides three convenient wrappers <code>adapt_glm</code>, <code>adapt_gam</code> and <code>adapt_glmnet</code>, which generate the results in Section 5 of <span class="citation">(Lei and Fithian 2016)</span>. In later versions we will add <code>adapt_gbm</code> (generalized boosting machine), <code>adapt_rf</code> (random forest), etc.. We recommend reading our paper first to familiarize with the basic concepts and scanning the documentation of them; See <code><a href="../reference/adapt_glm.html">?adapt_glm</a></code>, <code><a href="../reference/adapt_gam.html">?adapt_gam</a></code>, <code><a href="../reference/adapt_glmnet.html">?adapt_glmnet</a></code>.</p>
<div id="using-adapt_glm" class="section level2">
<h2 class="hasAnchor">
<a href="#using-adapt_glm" class="anchor"></a>Using <code>adapt_glm</code>
</h2>
<p>We illustrate one of the main function <code>adapt_glm</code>, for AdaPT with logistic-Gamma GLM as the working model, on <code>estrogen</code> dataset, a gene/drug response dataset from NCBI Gene Expression Omnibus (GEO). <code>estrogen</code> dataset consists of gene expression measurements for <span class="math inline">\(n = 22283\)</span> genes, in response to estrogen treatments in breast cancer cells for five groups of patients, with different dosage levels and 5 trials in each. The task is to identify the genes responding to a low dosage. The p-values pi for gene i is obtained by a one-sided permutation test which evaluates evidence for a change in gene expression level between the control group (placebo) and the low-dose group. <span class="math inline">\(\{p_i : i \in [n]\}\)</span> are then ordered according to permutation t-statistics comparing the control and low-dose data, pooled, against data from a higher dosage (with genes that appear to have a strong response at higher dosages placed earlier in the list). The code to compute p-values and the ordering can be found in <a href="http://www.stat.uchicago.edu/~rina/sabha/gene_drug_data_example.R">Rina Barber’s website</a>.</p>
<p>Here we subsample the top 5000 genes for illustration. The results for the full dataset are given in Section 5.1 of <span class="citation">(Lei and Fithian 2016)</span>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># load the data.</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"estrogen"</span>)</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co"># Take the first 5000 genes</span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"dplyr"</span>)){</span>
<span id="cb2-5"><a href="#cb2-5"></a>  <span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"dplyr"</span>)</span>
<span id="cb2-6"><a href="#cb2-6"></a>}</span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co">#&gt; Loading required namespace: dplyr</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"dplyr"</span>)</span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="co">#&gt; </span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="co">#&gt; </span></span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="co">#&gt;     filter, lag</span></span>
<span id="cb2-14"><a href="#cb2-14"></a><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="co">#&gt; </span></span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span id="cb2-17"><a href="#cb2-17"></a>estrogen &lt;-<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(estrogen, pvals, ord_high) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-18"><a href="#cb2-18"></a><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(ord_high <span class="op">&lt;=</span><span class="st"> </span><span class="dv">5000</span>)</span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(estrogen) &lt;-<span class="st"> </span><span class="ot">NULL</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(estrogen, <span class="dv">5</span>)</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="co">#&gt;        pvals ord_high</span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co">#&gt; 1 0.05011062      366</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co">#&gt; 2 0.71404053     4772</span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">#&gt; 3 0.06675970     3562</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co">#&gt; 4 0.40392007      790</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co">#&gt; 5 0.40614415     2291</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(estrogen)</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co">#&gt;      pvals             ord_high   </span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co">#&gt;  Min.   :0.000011   Min.   :   1  </span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co">#&gt;  1st Qu.:0.076082   1st Qu.:1251  </span></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="co">#&gt;  Median :0.238279   Median :2500  </span></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="co">#&gt;  Mean   :0.315094   Mean   :2500  </span></span>
<span id="cb4-7"><a href="#cb4-7"></a><span class="co">#&gt;  3rd Qu.:0.501009   3rd Qu.:3750  </span></span>
<span id="cb4-8"><a href="#cb4-8"></a><span class="co">#&gt;  Max.   :0.999289   Max.   :5000</span></span></code></pre></div>
<p><img src="adapt_demo_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
<p>Now we run <code>adapt_glm</code> on this dataset. <code>adapt_glm</code> takes a conditional logistic-Gamma GLM as the working model by default. Specifically, it models the p-values as <span class="math display">\[H_i \mid x_i \sim \pi(x_i), \quad \mathrm{logit}(\pi(x_i))= \phi(x_i)^{T}\beta\]</span> <span class="math display">\[-\log p_i \mid H_i, x_i\sim \left\{\begin{array}{ll} \mathrm{Exp(1)} &amp; H_i = 0\\ \mathrm{Exp(\mu(x))} &amp; H_i = 1\end{array}\right., \quad \frac{1}{\mu(x_i)} = \phi(x_i)^{T}\gamma\]</span> where <span class="math inline">\(\phi(x)\)</span> is a featurization of <span class="math inline">\(x\)</span>. In this example, we use the spline bases, given by <code>ns</code> function from <code>splines</code> package. For illustration, we choose our candidate models as the above GLMs with <span class="math inline">\(\phi(x)\)</span> being the spline bases with equal-spaced knots and the number of knots ranging from 6-10. We use BIC to select the best model at the initial stage and use the selected model for the following model fitting.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># prepare the inputs of AdaPT</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="co"># need "splines" package to construct the formula for glm</span></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"splines"</span>)</span>
<span id="cb5-4"><a href="#cb5-4"></a>pvals &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(estrogen<span class="op">$</span>pvals)</span>
<span id="cb5-5"><a href="#cb5-5"></a>x &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(estrogen<span class="op">$</span>ord))</span>
<span id="cb5-6"><a href="#cb5-6"></a>formulas &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"ns(x, df = "</span>, <span class="dv">6</span><span class="op">:</span><span class="dv">10</span>, <span class="st">")"</span>)</span>
<span id="cb5-7"><a href="#cb5-7"></a>formulas</span>
<span id="cb5-8"><a href="#cb5-8"></a><span class="co">#&gt; [1] "ns(x, df = 6)"  "ns(x, df = 7)"  "ns(x, df = 8)"  "ns(x, df = 9)" </span></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="co">#&gt; [5] "ns(x, df = 10)"</span></span></code></pre></div>
<p><code>adapt_glm</code> function provides several user-friendly tools to monitor the progress. For model selection, a progress bar will, by default, be shown in the console that indicates how much proportion of models have been fitted. This can be turned off by setting <code>verbose$ms = FALSE</code>. Similarly for model fitting, a progress bar can be shown in the console, though not by default, by setting <code>verbose$fit = TRUE</code>. Also, by default, the progress of the main process will be shown in the console that indicates (1) which target FDR level has been achieved; (2) FDPhat for each target FDR level; (3) number of rejections for each target FDR level.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="co"># run AdaPT</span></span>
<span id="cb6-2"><a href="#cb6-2"></a>res_glm &lt;-<span class="st"> </span><span class="kw"><a href="../reference/adapt_glm.html">adapt_glm</a></span>(<span class="dt">x =</span> x, <span class="dt">pvals =</span> pvals, <span class="dt">pi_formulas =</span> formulas, <span class="dt">mu_formulas =</span> formulas)</span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co">#&gt; Model selection starts!</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co">#&gt; Shrink the set of candidate models if it is too time-consuming.</span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">#&gt; </span></span>
<span id="cb6-6"><a href="#cb6-6"></a>  <span class="op">|</span><span class="st">                                                        </span></span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="st">  </span><span class="er">|</span><span class="st">                                                  </span><span class="er">|</span><span class="st">   </span><span class="dv">0</span>%</span>
<span id="cb6-8"><a href="#cb6-8"></a>  <span class="op">|</span><span class="st">                                                        </span></span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="st">  </span><span class="er">|==========</span><span class="st">                                        </span><span class="er">|</span><span class="st">  </span><span class="dv">20</span>%</span>
<span id="cb6-10"><a href="#cb6-10"></a>  <span class="op">|</span><span class="st">                                                        </span></span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="st">  </span><span class="er">|====================</span><span class="st">                              </span><span class="er">|</span><span class="st">  </span><span class="dv">40</span>%</span>
<span id="cb6-12"><a href="#cb6-12"></a>  <span class="op">|</span><span class="st">                                                        </span></span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="st">  </span><span class="er">|==============================</span><span class="st">                    </span><span class="er">|</span><span class="st">  </span><span class="dv">60</span>%</span>
<span id="cb6-14"><a href="#cb6-14"></a>  <span class="op">|</span><span class="st">                                                        </span></span>
<span id="cb6-15"><a href="#cb6-15"></a><span class="st">  </span><span class="er">|========================================</span><span class="st">          </span><span class="er">|</span><span class="st">  </span><span class="dv">80</span>%</span>
<span id="cb6-16"><a href="#cb6-16"></a>  <span class="op">|</span><span class="st">                                                        </span></span>
<span id="cb6-17"><a href="#cb6-17"></a><span class="st">  </span><span class="er">|==================================================|</span><span class="st"> </span><span class="dv">100</span>%</span>
<span id="cb6-18"><a href="#cb6-18"></a><span class="co">#&gt; alpha = 0.29: FDPhat 0.2899, Number of Rej. 3474</span></span>
<span id="cb6-19"><a href="#cb6-19"></a><span class="co">#&gt; alpha = 0.28: FDPhat 0.28, Number of Rej. 3325</span></span>
<span id="cb6-20"><a href="#cb6-20"></a><span class="co">#&gt; alpha = 0.27: FDPhat 0.2698, Number of Rej. 3084</span></span>
<span id="cb6-21"><a href="#cb6-21"></a><span class="co">#&gt; alpha = 0.26: FDPhat 0.2598, Number of Rej. 2925</span></span>
<span id="cb6-22"><a href="#cb6-22"></a><span class="co">#&gt; alpha = 0.25: FDPhat 0.2498, Number of Rej. 2882</span></span>
<span id="cb6-23"><a href="#cb6-23"></a><span class="co">#&gt; alpha = 0.24: FDPhat 0.2399, Number of Rej. 2834</span></span>
<span id="cb6-24"><a href="#cb6-24"></a><span class="co">#&gt; alpha = 0.23: FDPhat 0.23, Number of Rej. 2761</span></span>
<span id="cb6-25"><a href="#cb6-25"></a><span class="co">#&gt; alpha = 0.22: FDPhat 0.2198, Number of Rej. 2703</span></span>
<span id="cb6-26"><a href="#cb6-26"></a><span class="co">#&gt; alpha = 0.21: FDPhat 0.2096, Number of Rej. 2595</span></span>
<span id="cb6-27"><a href="#cb6-27"></a><span class="co">#&gt; alpha = 0.2: FDPhat 0.1998, Number of Rej. 2547</span></span>
<span id="cb6-28"><a href="#cb6-28"></a><span class="co">#&gt; alpha = 0.19: FDPhat 0.1897, Number of Rej. 2398</span></span>
<span id="cb6-29"><a href="#cb6-29"></a><span class="co">#&gt; alpha = 0.18: FDPhat 0.1797, Number of Rej. 2287</span></span>
<span id="cb6-30"><a href="#cb6-30"></a><span class="co">#&gt; alpha = 0.17: FDPhat 0.1699, Number of Rej. 2231</span></span>
<span id="cb6-31"><a href="#cb6-31"></a><span class="co">#&gt; alpha = 0.16: FDPhat 0.1599, Number of Rej. 2139</span></span>
<span id="cb6-32"><a href="#cb6-32"></a><span class="co">#&gt; alpha = 0.15: FDPhat 0.15, Number of Rej. 2060</span></span>
<span id="cb6-33"><a href="#cb6-33"></a><span class="co">#&gt; alpha = 0.14: FDPhat 0.1399, Number of Rej. 1966</span></span>
<span id="cb6-34"><a href="#cb6-34"></a><span class="co">#&gt; alpha = 0.13: FDPhat 0.1297, Number of Rej. 1889</span></span>
<span id="cb6-35"><a href="#cb6-35"></a><span class="co">#&gt; alpha = 0.12: FDPhat 0.1195, Number of Rej. 1824</span></span>
<span id="cb6-36"><a href="#cb6-36"></a><span class="co">#&gt; alpha = 0.11: FDPhat 0.1097, Number of Rej. 1723</span></span>
<span id="cb6-37"><a href="#cb6-37"></a><span class="co">#&gt; alpha = 0.1: FDPhat 0.0999, Number of Rej. 1582</span></span>
<span id="cb6-38"><a href="#cb6-38"></a><span class="co">#&gt; alpha = 0.09: FDPhat 0.0895, Number of Rej. 1430</span></span>
<span id="cb6-39"><a href="#cb6-39"></a><span class="co">#&gt; alpha = 0.08: FDPhat 0.0798, Number of Rej. 1240</span></span>
<span id="cb6-40"><a href="#cb6-40"></a><span class="co">#&gt; alpha = 0.07: FDPhat 0.0697, Number of Rej. 1105</span></span>
<span id="cb6-41"><a href="#cb6-41"></a><span class="co">#&gt; alpha = 0.06: FDPhat 0.0594, Number of Rej. 960</span></span>
<span id="cb6-42"><a href="#cb6-42"></a><span class="co">#&gt; alpha = 0.05: FDPhat 0.0495, Number of Rej. 909</span></span>
<span id="cb6-43"><a href="#cb6-43"></a><span class="co">#&gt; alpha = 0.04: FDPhat 0.0389, Number of Rej. 796</span></span>
<span id="cb6-44"><a href="#cb6-44"></a><span class="co">#&gt; alpha = 0.03: FDPhat 0.0297, Number of Rej. 539</span></span>
<span id="cb6-45"><a href="#cb6-45"></a><span class="co">#&gt; alpha = 0.02: FDPhat 0.0179, Number of Rej. 224</span></span></code></pre></div>
<p><code>plot_1d_thresh</code> gives the plot for the rejection threshold as a function of x (must be univariate without repeated value) for given <span class="math inline">\(\alpha\)</span>. We display the plots for <span class="math inline">\(\alpha \in \{0.3, 0.25, 0.2, 0.15, 0.1, 0.05\}\)</span>.</p>
<p><img src="adapt_demo_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<p><code>plot_1d_lfdr</code> gives the plot for the estimated local FDR as a function of x (must be univariate without repeated value) for given <span class="math inline">\(\alpha\)</span>. We display the plots for <span class="math inline">\(\alpha \in \{0.3, 0.25, 0.2, 0.15, 0.1, 0.05\}\)</span>. It is clear that the estimated local FDR almost remains the same, indicating that the information loss caused by partial masking is small.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="dt">mfrow =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">2</span>, <span class="dv">3</span>))</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="cf">for</span> (alpha <span class="cf">in</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">0.3</span>, <span class="fl">0.05</span>, <span class="fl">-0.05</span>)){</span>
<span id="cb7-3"><a href="#cb7-3"></a>  nrejs &lt;-<span class="st"> </span>res_glm<span class="op">$</span>nrejs[<span class="kw"><a href="https://rdrr.io/r/base/Round.html">floor</a></span>(alpha <span class="op">*</span><span class="st"> </span><span class="dv">100</span>)]</span>
<span id="cb7-4"><a href="#cb7-4"></a>  title &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"alpha = "</span>, alpha, <span class="st">", nrejs = "</span>, nrejs)</span>
<span id="cb7-5"><a href="#cb7-5"></a>  <span class="kw"><a href="../reference/plot_1d.html">plot_1d_lfdr</a></span>(res_glm, x, pvals, alpha, title, <span class="dt">disp_ymax =</span> <span class="fl">0.25</span>, <span class="dt">xlab =</span> <span class="st">"order"</span>)</span>
<span id="cb7-6"><a href="#cb7-6"></a>}</span></code></pre></div>
<p><img src="adapt_demo_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
</div>
<div id="using-adapt_gam" class="section level2">
<h2 class="hasAnchor">
<a href="#using-adapt_gam" class="anchor"></a>Using <code>adapt_gam</code>
</h2>
<p><code>adapt_gam</code> has the exactly the same setting as <code>adapt_glm</code> except that a generalized additive model (GAM) is fitted instead of a generalized linear model. We refer the readers to <span class="citation">(Wood 2006)</span> for more details on GAMs.</p>
<p>We illustrate <code>adapt_gam</code> using a simulation dataset, under similar settings in Section 5.2 of <span class="citation">(Lei and Fithian 2016)</span>. In this case, <span class="math inline">\(x_{i}\)</span> is a two dimensional vector generated from an equi-spaced <span class="math inline">\(20\times 20\)</span> grid in the area <span class="math inline">\([-100, 100]\times [-100, 100]\)</span>. The p-values are generated from a one-sided normal test, i.e. <span class="math display">\[p_i = 1 - \Phi(z_i), \quad z_i\sim N(\mu_i, 1), \quad \mu_i = \left\{\begin{array}{cc}0 &amp; (i\in \mathcal{H}_{0})\\
2 &amp; (i\not\in \mathcal{H}_{0})\end{array}\right.\]</span> We set the hypotheses in a circle in the center of the grid with radius 40. The following code generates the p-values:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># Generate a 2-dim x</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>n &lt;-<span class="st"> </span><span class="dv">400</span></span>
<span id="cb8-3"><a href="#cb8-3"></a>x1 &lt;-<span class="st"> </span>x2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="op">-</span><span class="dv">100</span>, <span class="dv">100</span>, <span class="dt">length.out =</span> <span class="dv">20</span>)</span>
<span id="cb8-4"><a href="#cb8-4"></a>x &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/expand.grid.html">expand.grid</a></span>(x1, x2)</span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(x) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"x1"</span>, <span class="st">"x2"</span>)</span>
<span id="cb8-6"><a href="#cb8-6"></a></span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="co"># Generate p-values (one-sided z test)</span></span>
<span id="cb8-8"><a href="#cb8-8"></a>H0 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(x, <span class="dv">1</span>, <span class="cf">function</span>(coord){<span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(coord<span class="op">^</span><span class="dv">2</span>) <span class="op">&lt;</span><span class="st"> </span><span class="dv">40</span><span class="op">^</span><span class="dv">2</span>})</span>
<span id="cb8-9"><a href="#cb8-9"></a>mu &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(H0, <span class="dv">2</span>, <span class="dv">0</span>)</span>
<span id="cb8-10"><a href="#cb8-10"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">0</span>)</span>
<span id="cb8-11"><a href="#cb8-11"></a>zvals &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(n) <span class="op">+</span><span class="st"> </span>mu</span>
<span id="cb8-12"><a href="#cb8-12"></a>pvals &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">pnorm</a></span>(zvals)</span></code></pre></div>
<p>Now we apply <code>adapt_gam</code> on this dataset. <code>adapt_gam</code> is built on <code>mgcv</code> package which provides a more advanced version of <code>gam</code> in <code>stats</code> package. Users need to install <code>mgcv</code> package first.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co"># install.packages("mgcv")</span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"mgcv"</span>)</span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="co">#&gt; Loading required package: nlme</span></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="co">#&gt; </span></span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="co">#&gt; Attaching package: 'nlme'</span></span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="co">#&gt; </span></span>
<span id="cb9-8"><a href="#cb9-8"></a><span class="co">#&gt;     collapse</span></span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="co">#&gt; This is mgcv 1.8-31. For overview type 'help("mgcv-package")'.</span></span></code></pre></div>
<p>By default, <code>adapt_gam</code> takes a logistic-Gamma GAM as the working model and the only modelling input is the (list of) partial formula(s) as specified in <code>adapt_glm</code>. Here we take advantage of automatic knots selection of function <code>s</code>, based on generalized cross-validation criterion, in <code>mgcv</code> function to get rid of manual model selection; See <code><a href="https://rdrr.io/pkg/mgcv/man/s.html">?s</a></code> for details. In particular, we consider a two-dimensional spline basis for <span class="math inline">\(x_1\)</span> and <span class="math inline">\(x_2\)</span>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>formula &lt;-<span class="st"> "s(x1, x2)"</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>res_gam &lt;-<span class="st"> </span><span class="kw"><a href="../reference/adapt_gam.html">adapt_gam</a></span>(<span class="dt">x =</span> x, <span class="dt">pvals =</span> pvals, <span class="dt">pi_formulas =</span> formula, <span class="dt">mu_formulas =</span> formula)</span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="co">#&gt; alpha = 0.83: FDPhat 0.8299, Number of Rej. 194</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="co">#&gt; alpha = 0.82: FDPhat 0.8196, Number of Rej. 194</span></span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="co">#&gt; alpha = 0.81: FDPhat 0.8093, Number of Rej. 194</span></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="co">#&gt; alpha = 0.8: FDPhat 0.7979, Number of Rej. 193</span></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="co">#&gt; alpha = 0.79: FDPhat 0.7895, Number of Rej. 190</span></span>
<span id="cb10-8"><a href="#cb10-8"></a><span class="co">#&gt; alpha = 0.78: FDPhat 0.7789, Number of Rej. 190</span></span>
<span id="cb10-9"><a href="#cb10-9"></a><span class="co">#&gt; alpha = 0.77: FDPhat 0.7684, Number of Rej. 190</span></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="co">#&gt; alpha = 0.76: FDPhat 0.7579, Number of Rej. 190</span></span>
<span id="cb10-11"><a href="#cb10-11"></a><span class="co">#&gt; alpha = 0.75: FDPhat 0.75, Number of Rej. 172</span></span>
<span id="cb10-12"><a href="#cb10-12"></a><span class="co">#&gt; alpha = 0.74: FDPhat 0.7349, Number of Rej. 166</span></span>
<span id="cb10-13"><a href="#cb10-13"></a><span class="co">#&gt; alpha = 0.73: FDPhat 0.7273, Number of Rej. 165</span></span>
<span id="cb10-14"><a href="#cb10-14"></a><span class="co">#&gt; alpha = 0.72: FDPhat 0.7195, Number of Rej. 164</span></span>
<span id="cb10-15"><a href="#cb10-15"></a><span class="co">#&gt; alpha = 0.71: FDPhat 0.7055, Number of Rej. 163</span></span>
<span id="cb10-16"><a href="#cb10-16"></a><span class="co">#&gt; alpha = 0.7: FDPhat 0.6987, Number of Rej. 156</span></span>
<span id="cb10-17"><a href="#cb10-17"></a><span class="co">#&gt; alpha = 0.69: FDPhat 0.6842, Number of Rej. 152</span></span>
<span id="cb10-18"><a href="#cb10-18"></a><span class="co">#&gt; alpha = 0.68: FDPhat 0.6776, Number of Rej. 152</span></span>
<span id="cb10-19"><a href="#cb10-19"></a><span class="co">#&gt; alpha = 0.67: FDPhat 0.6667, Number of Rej. 150</span></span>
<span id="cb10-20"><a href="#cb10-20"></a><span class="co">#&gt; alpha = 0.66: FDPhat 0.66, Number of Rej. 150</span></span>
<span id="cb10-21"><a href="#cb10-21"></a><span class="co">#&gt; alpha = 0.65: FDPhat 0.6467, Number of Rej. 150</span></span>
<span id="cb10-22"><a href="#cb10-22"></a><span class="co">#&gt; alpha = 0.64: FDPhat 0.64, Number of Rej. 150</span></span>
<span id="cb10-23"><a href="#cb10-23"></a><span class="co">#&gt; alpha = 0.63: FDPhat 0.627, Number of Rej. 126</span></span>
<span id="cb10-24"><a href="#cb10-24"></a><span class="co">#&gt; alpha = 0.62: FDPhat 0.6129, Number of Rej. 124</span></span>
<span id="cb10-25"><a href="#cb10-25"></a><span class="co">#&gt; alpha = 0.61: FDPhat 0.6048, Number of Rej. 124</span></span>
<span id="cb10-26"><a href="#cb10-26"></a><span class="co">#&gt; alpha = 0.6: FDPhat 0.5968, Number of Rej. 124</span></span>
<span id="cb10-27"><a href="#cb10-27"></a><span class="co">#&gt; alpha = 0.59: FDPhat 0.5854, Number of Rej. 123</span></span>
<span id="cb10-28"><a href="#cb10-28"></a><span class="co">#&gt; alpha = 0.58: FDPhat 0.5772, Number of Rej. 123</span></span>
<span id="cb10-29"><a href="#cb10-29"></a><span class="co">#&gt; alpha = 0.57: FDPhat 0.5691, Number of Rej. 123</span></span>
<span id="cb10-30"><a href="#cb10-30"></a><span class="co">#&gt; alpha = 0.56: FDPhat 0.5574, Number of Rej. 122</span></span>
<span id="cb10-31"><a href="#cb10-31"></a><span class="co">#&gt; alpha = 0.55: FDPhat 0.5492, Number of Rej. 122</span></span>
<span id="cb10-32"><a href="#cb10-32"></a><span class="co">#&gt; alpha = 0.54: FDPhat 0.5372, Number of Rej. 121</span></span>
<span id="cb10-33"><a href="#cb10-33"></a><span class="co">#&gt; alpha = 0.53: FDPhat 0.5254, Number of Rej. 118</span></span>
<span id="cb10-34"><a href="#cb10-34"></a><span class="co">#&gt; alpha = 0.52: FDPhat 0.5133, Number of Rej. 113</span></span>
<span id="cb10-35"><a href="#cb10-35"></a><span class="co">#&gt; alpha = 0.51: FDPhat 0.5046, Number of Rej. 109</span></span>
<span id="cb10-36"><a href="#cb10-36"></a><span class="co">#&gt; alpha = 0.5: FDPhat 0.4948, Number of Rej. 97</span></span>
<span id="cb10-37"><a href="#cb10-37"></a><span class="co">#&gt; alpha = 0.49: FDPhat 0.4845, Number of Rej. 97</span></span>
<span id="cb10-38"><a href="#cb10-38"></a><span class="co">#&gt; alpha = 0.48: FDPhat 0.4787, Number of Rej. 94</span></span>
<span id="cb10-39"><a href="#cb10-39"></a><span class="co">#&gt; alpha = 0.47: FDPhat 0.4674, Number of Rej. 92</span></span>
<span id="cb10-40"><a href="#cb10-40"></a><span class="co">#&gt; alpha = 0.46: FDPhat 0.4565, Number of Rej. 92</span></span>
<span id="cb10-41"><a href="#cb10-41"></a><span class="co">#&gt; alpha = 0.45: FDPhat 0.4396, Number of Rej. 91</span></span>
<span id="cb10-42"><a href="#cb10-42"></a><span class="co">#&gt; alpha = 0.44: FDPhat 0.4396, Number of Rej. 91</span></span>
<span id="cb10-43"><a href="#cb10-43"></a><span class="co">#&gt; alpha = 0.43: FDPhat 0.4286, Number of Rej. 91</span></span>
<span id="cb10-44"><a href="#cb10-44"></a><span class="co">#&gt; alpha = 0.42: FDPhat 0.4176, Number of Rej. 91</span></span>
<span id="cb10-45"><a href="#cb10-45"></a><span class="co">#&gt; alpha = 0.41: FDPhat 0.4066, Number of Rej. 91</span></span>
<span id="cb10-46"><a href="#cb10-46"></a><span class="co">#&gt; alpha = 0.4: FDPhat 0.3924, Number of Rej. 79</span></span>
<span id="cb10-47"><a href="#cb10-47"></a><span class="co">#&gt; alpha = 0.39: FDPhat 0.3846, Number of Rej. 78</span></span>
<span id="cb10-48"><a href="#cb10-48"></a><span class="co">#&gt; alpha = 0.38: FDPhat 0.3718, Number of Rej. 78</span></span>
<span id="cb10-49"><a href="#cb10-49"></a><span class="co">#&gt; alpha = 0.37: FDPhat 0.3611, Number of Rej. 72</span></span>
<span id="cb10-50"><a href="#cb10-50"></a><span class="co">#&gt; alpha = 0.36: FDPhat 0.3472, Number of Rej. 72</span></span>
<span id="cb10-51"><a href="#cb10-51"></a><span class="co">#&gt; alpha = 0.35: FDPhat 0.3472, Number of Rej. 72</span></span>
<span id="cb10-52"><a href="#cb10-52"></a><span class="co">#&gt; alpha = 0.34: FDPhat 0.338, Number of Rej. 71</span></span>
<span id="cb10-53"><a href="#cb10-53"></a><span class="co">#&gt; alpha = 0.33: FDPhat 0.3286, Number of Rej. 70</span></span>
<span id="cb10-54"><a href="#cb10-54"></a><span class="co">#&gt; alpha = 0.32: FDPhat 0.3188, Number of Rej. 69</span></span>
<span id="cb10-55"><a href="#cb10-55"></a><span class="co">#&gt; alpha = 0.31: FDPhat 0.3043, Number of Rej. 69</span></span>
<span id="cb10-56"><a href="#cb10-56"></a><span class="co">#&gt; alpha = 0.3: FDPhat 0.2899, Number of Rej. 69</span></span>
<span id="cb10-57"><a href="#cb10-57"></a><span class="co">#&gt; alpha = 0.29: FDPhat 0.2899, Number of Rej. 69</span></span>
<span id="cb10-58"><a href="#cb10-58"></a><span class="co">#&gt; alpha = 0.28: FDPhat 0.2754, Number of Rej. 69</span></span>
<span id="cb10-59"><a href="#cb10-59"></a><span class="co">#&gt; alpha = 0.27: FDPhat 0.2647, Number of Rej. 68</span></span>
<span id="cb10-60"><a href="#cb10-60"></a><span class="co">#&gt; alpha = 0.26: FDPhat 0.2576, Number of Rej. 66</span></span>
<span id="cb10-61"><a href="#cb10-61"></a><span class="co">#&gt; alpha = 0.25: FDPhat 0.2462, Number of Rej. 65</span></span>
<span id="cb10-62"><a href="#cb10-62"></a><span class="co">#&gt; alpha = 0.24: FDPhat 0.2381, Number of Rej. 63</span></span>
<span id="cb10-63"><a href="#cb10-63"></a><span class="co">#&gt; alpha = 0.23: FDPhat 0.2241, Number of Rej. 58</span></span>
<span id="cb10-64"><a href="#cb10-64"></a><span class="co">#&gt; alpha = 0.22: FDPhat 0.2069, Number of Rej. 58</span></span>
<span id="cb10-65"><a href="#cb10-65"></a><span class="co">#&gt; alpha = 0.21: FDPhat 0.2069, Number of Rej. 58</span></span>
<span id="cb10-66"><a href="#cb10-66"></a><span class="co">#&gt; alpha = 0.2: FDPhat 0.193, Number of Rej. 57</span></span>
<span id="cb10-67"><a href="#cb10-67"></a><span class="co">#&gt; alpha = 0.19: FDPhat 0.1786, Number of Rej. 56</span></span>
<span id="cb10-68"><a href="#cb10-68"></a><span class="co">#&gt; alpha = 0.18: FDPhat 0.1786, Number of Rej. 56</span></span>
<span id="cb10-69"><a href="#cb10-69"></a><span class="co">#&gt; alpha = 0.17: FDPhat 0.1667, Number of Rej. 54</span></span>
<span id="cb10-70"><a href="#cb10-70"></a><span class="co">#&gt; alpha = 0.16: FDPhat 0.1481, Number of Rej. 54</span></span>
<span id="cb10-71"><a href="#cb10-71"></a><span class="co">#&gt; alpha = 0.15: FDPhat 0.1481, Number of Rej. 54</span></span>
<span id="cb10-72"><a href="#cb10-72"></a><span class="co">#&gt; alpha = 0.14: FDPhat 0.1296, Number of Rej. 54</span></span>
<span id="cb10-73"><a href="#cb10-73"></a><span class="co">#&gt; alpha = 0.13: FDPhat 0.1296, Number of Rej. 54</span></span>
<span id="cb10-74"><a href="#cb10-74"></a><span class="co">#&gt; alpha = 0.12: FDPhat 0.1154, Number of Rej. 52</span></span>
<span id="cb10-75"><a href="#cb10-75"></a><span class="co">#&gt; alpha = 0.11: FDPhat 0.098, Number of Rej. 51</span></span>
<span id="cb10-76"><a href="#cb10-76"></a><span class="co">#&gt; alpha = 0.1: FDPhat 0.098, Number of Rej. 51</span></span>
<span id="cb10-77"><a href="#cb10-77"></a><span class="co">#&gt; alpha = 0.09: FDPhat 0.0833, Number of Rej. 48</span></span>
<span id="cb10-78"><a href="#cb10-78"></a><span class="co">#&gt; alpha = 0.08: FDPhat 0.0789, Number of Rej. 38</span></span>
<span id="cb10-79"><a href="#cb10-79"></a><span class="co">#&gt; alpha = 0.07: FDPhat 0.0571, Number of Rej. 35</span></span>
<span id="cb10-80"><a href="#cb10-80"></a><span class="co">#&gt; alpha = 0.06: FDPhat 0.0571, Number of Rej. 35</span></span></code></pre></div>
<p>We visualize the results below. The left figure gives the truth where the darker pixels correspond to non-nulls and the right figure gives the rejection set by <code>adapt_gam</code> for <span class="math inline">\(\alpha = 0.1\)</span>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="dt">mfrow =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="dt">mar =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">4</span>, <span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="co"># Truth</span></span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(x[, <span class="dv">1</span>], x[, <span class="dv">2</span>], <span class="dt">type =</span> <span class="st">"p"</span>, <span class="dt">xlab =</span> <span class="st">""</span>, <span class="dt">ylab =</span> <span class="st">""</span>, <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(H0, <span class="st">"#000000"</span>, <span class="st">"#A9A9A9"</span>), <span class="dt">cex =</span> <span class="dv">2</span>, <span class="dt">pch =</span> <span class="dv">15</span>)</span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="co"># Rejection set</span></span>
<span id="cb11-5"><a href="#cb11-5"></a>rej &lt;-<span class="st"> </span>pvals <span class="op">&lt;=</span><span class="st"> </span>res_gam<span class="op">$</span>s[, <span class="dv">10</span>]</span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(x[, <span class="dv">1</span>], x[, <span class="dv">2</span>], <span class="dt">type =</span> <span class="st">"p"</span>, <span class="dt">xlab =</span> <span class="st">""</span>, <span class="dt">ylab =</span> <span class="st">""</span>, <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(rej, <span class="st">"#800000"</span>, <span class="st">"#FFB6C1"</span>), <span class="dt">cex =</span> <span class="dv">2</span>, <span class="dt">pch =</span> <span class="dv">15</span>)</span></code></pre></div>
<p><img src="adapt_demo_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<p>Similar to the 1d case, <code>plot_2d_thresh</code> gives the plot for the rejection threshold as a function of x (must be bivariate without repeated value) for given <span class="math inline">\(\alpha\)</span>. In particular we set <span class="math inline">\(\alpha = 0.1\)</span>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="kw"><a href="../reference/plot_2d.html">plot_2d_thresh</a></span>(res_gam, x, pvals, <span class="fl">0.1</span>, <span class="st">"P-Value Threshold"</span>, <span class="dt">xlab =</span> <span class="st">""</span>, <span class="dt">ylab =</span> <span class="st">""</span>)</span></code></pre></div>
<p><img src="adapt_demo_files/figure-html/unnamed-chunk-15-1.png" width="576"></p>
<p>Unlike the 1d case, visualizing the level surfaces of local FDR estimates is not an easy task due to the extra dimension. <code>plot_2d_lfdr</code> gives the plot for the local FDR estimates when all p-values are equal to a user-specified level <code>targetp</code> for given <span class="math inline">\(\alpha\)</span>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="kw"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="dt">mfrow =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">2</span>, <span class="dv">2</span>), <span class="dt">mar =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">4</span>, <span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="cf">for</span> (targetp <span class="cf">in</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.001</span>, <span class="fl">0.005</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span>)){</span>
<span id="cb13-3"><a href="#cb13-3"></a>  title &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"Local FDR Estimates (p = "</span>, targetp, <span class="st">")"</span>)</span>
<span id="cb13-4"><a href="#cb13-4"></a>  <span class="kw"><a href="../reference/plot_2d.html">plot_2d_lfdr</a></span>(res_gam, x, pvals, <span class="fl">0.1</span>, title, targetp, <span class="dt">xlab =</span> <span class="st">""</span>, <span class="dt">ylab =</span> <span class="st">""</span>)</span>
<span id="cb13-5"><a href="#cb13-5"></a>}</span></code></pre></div>
<p><img src="adapt_demo_files/figure-html/unnamed-chunk-16-1.png" width="576"><img src="adapt_demo_files/figure-html/unnamed-chunk-16-2.png" width="576"><img src="adapt_demo_files/figure-html/unnamed-chunk-16-3.png" width="576"><img src="adapt_demo_files/figure-html/unnamed-chunk-16-4.png" width="576"></p>
</div>
<div id="using-adapt_glmnet" class="section level2">
<h2 class="hasAnchor">
<a href="#using-adapt_glmnet" class="anchor"></a>Using <code>adapt_glmnet</code>
</h2>
<p><code>adapt_glmnet</code> is appropriate for the cases where the covariates are high-dimensional. Instead of fitting GLMs and GAMs, <code>adapt_glmnet</code> fits a penalized GLM in each step. Unlike <code>adapt_glm</code> and <code>adapt_gam</code>, the covariate <span class="math inline">\(x\)</span> should be of class <code>matrix</code>, instead of <code>data.frame</code>, and <code>adapt_glmnet</code> always take the full matrix <span class="math inline">\(x\)</span> as the covariates. As a result, <code>adapt_glmnet</code> has a simpler interface than <code>adapt_glm</code> and <code>adapt_gam</code> in that it does not need inputs like <code>formula</code>. We refer the readers to <span class="citation">(Hastie and Qian 2014)</span> for more details on <code>glmnet</code>.</p>
<p>We illustrate <code>adapt_glmnet</code> using a simulation dataset, under similar settings in Section 5.2 of <span class="citation">(Lei and Fithian 2016)</span>. In this case, we generate 300 i.i.d. <span class="math inline">\(x_{i}\)</span>’s where each of them is a <span class="math inline">\(50\)</span>-dimensional vector with each entry generated from an uniform distribution on <span class="math inline">\([0, 1]\)</span>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">0</span>)</span>
<span id="cb14-2"><a href="#cb14-2"></a>m &lt;-<span class="st"> </span><span class="dv">50</span></span>
<span id="cb14-3"><a href="#cb14-3"></a>n &lt;-<span class="st"> </span><span class="dv">500</span></span>
<span id="cb14-4"><a href="#cb14-4"></a>x &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(n <span class="op">*</span><span class="st"> </span>m), n, m)</span></code></pre></div>
<p>We consider the p-values from a logistic Gamma GLM model with <span class="math display">\[\pi_{1}(x_i) = \frac{\exp(-3 + 2x_{i1} + 2x_{i2})}{1 + \exp(-3 + 2x_{i1} + 2x_{i2})}, \quad \mu(x_i) = \max\{1, 2x_{i1} + 2x_{i2}\}\]</span></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>inv_logit &lt;-<span class="st"> </span><span class="cf">function</span>(x) {<span class="kw"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(x) <span class="op">/</span><span class="st"> </span>(<span class="dv">1</span> <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(x))}</span>
<span id="cb15-2"><a href="#cb15-2"></a>pi &lt;-<span class="st"> </span><span class="kw">inv_logit</span>(x[, <span class="dv">1</span>] <span class="op">*</span><span class="st"> </span><span class="dv">2</span> <span class="op">+</span><span class="st"> </span>x[, <span class="dv">2</span>] <span class="op">*</span><span class="st"> </span><span class="dv">2</span> <span class="op">-</span><span class="st"> </span><span class="dv">3</span>)</span>
<span id="cb15-3"><a href="#cb15-3"></a>mu &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">pmax</a></span>(<span class="dv">1</span>, x[, <span class="dv">1</span>] <span class="op">*</span><span class="st"> </span><span class="dv">2</span> <span class="op">+</span><span class="st"> </span>x[, <span class="dv">2</span>] <span class="op">*</span><span class="st"> </span><span class="dv">2</span>)</span></code></pre></div>
<p>Now we generate the null set and the p-values as follows:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>H0 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/logical.html">as.logical</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(n) <span class="op">&lt;</span><span class="st"> </span>pi, <span class="dv">1</span>, <span class="dv">0</span>))</span>
<span id="cb16-2"><a href="#cb16-2"></a>y &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(H0, <span class="kw"><a href="https://rdrr.io/r/stats/Exponential.html">rexp</a></span>(n, <span class="dv">1</span><span class="op">/</span>mu), <span class="kw"><a href="https://rdrr.io/r/stats/Exponential.html">rexp</a></span>(n, <span class="dv">1</span>))</span>
<span id="cb16-3"><a href="#cb16-3"></a>pvals &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(<span class="op">-</span>y)</span></code></pre></div>
<p>To save computation time, we start <code>adapt_glmnet</code> with <span class="math inline">\(s_{0} = (0.15, 0.15, \ldots, 0.15)\)</span>, instead of the default value <span class="math inline">\(s_{0} = (0.45, 0.45, \ldots, 0.45)\)</span>, and set <code>nfits = 5</code> to reduce the number of fitting in the process (see <code><a href="../reference/adapt.html">?adapt</a></code> for more details).</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/adapt_glmnet.html">adapt_glmnet</a></span>(x, pvals, <span class="dt">s0 =</span> <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">0.15</span>, n), <span class="dt">nfits =</span> <span class="dv">5</span>)</span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="co">#&gt; alpha = 0.59: FDPhat 0.5847, Number of Rej. 118</span></span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="co">#&gt; alpha = 0.58: FDPhat 0.5763, Number of Rej. 118</span></span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="co">#&gt; alpha = 0.57: FDPhat 0.5641, Number of Rej. 117</span></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="co">#&gt; alpha = 0.56: FDPhat 0.5556, Number of Rej. 117</span></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="co">#&gt; alpha = 0.55: FDPhat 0.547, Number of Rej. 117</span></span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="co">#&gt; alpha = 0.54: FDPhat 0.5345, Number of Rej. 116</span></span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="co">#&gt; alpha = 0.53: FDPhat 0.5263, Number of Rej. 114</span></span>
<span id="cb17-9"><a href="#cb17-9"></a><span class="co">#&gt; alpha = 0.52: FDPhat 0.5179, Number of Rej. 112</span></span>
<span id="cb17-10"><a href="#cb17-10"></a><span class="co">#&gt; alpha = 0.51: FDPhat 0.5, Number of Rej. 96</span></span>
<span id="cb17-11"><a href="#cb17-11"></a><span class="co">#&gt; alpha = 0.5: FDPhat 0.5, Number of Rej. 96</span></span>
<span id="cb17-12"><a href="#cb17-12"></a><span class="co">#&gt; alpha = 0.49: FDPhat 0.4896, Number of Rej. 96</span></span>
<span id="cb17-13"><a href="#cb17-13"></a><span class="co">#&gt; alpha = 0.48: FDPhat 0.4737, Number of Rej. 95</span></span>
<span id="cb17-14"><a href="#cb17-14"></a><span class="co">#&gt; alpha = 0.47: FDPhat 0.4632, Number of Rej. 95</span></span>
<span id="cb17-15"><a href="#cb17-15"></a><span class="co">#&gt; alpha = 0.46: FDPhat 0.4574, Number of Rej. 94</span></span>
<span id="cb17-16"><a href="#cb17-16"></a><span class="co">#&gt; alpha = 0.45: FDPhat 0.4457, Number of Rej. 92</span></span>
<span id="cb17-17"><a href="#cb17-17"></a><span class="co">#&gt; alpha = 0.44: FDPhat 0.4396, Number of Rej. 91</span></span>
<span id="cb17-18"><a href="#cb17-18"></a><span class="co">#&gt; alpha = 0.43: FDPhat 0.4286, Number of Rej. 91</span></span>
<span id="cb17-19"><a href="#cb17-19"></a><span class="co">#&gt; alpha = 0.42: FDPhat 0.4167, Number of Rej. 84</span></span>
<span id="cb17-20"><a href="#cb17-20"></a><span class="co">#&gt; alpha = 0.41: FDPhat 0.4074, Number of Rej. 81</span></span>
<span id="cb17-21"><a href="#cb17-21"></a><span class="co">#&gt; alpha = 0.4: FDPhat 0.3947, Number of Rej. 76</span></span>
<span id="cb17-22"><a href="#cb17-22"></a><span class="co">#&gt; alpha = 0.39: FDPhat 0.3889, Number of Rej. 54</span></span>
<span id="cb17-23"><a href="#cb17-23"></a><span class="co">#&gt; alpha = 0.38: FDPhat 0.3774, Number of Rej. 53</span></span>
<span id="cb17-24"><a href="#cb17-24"></a><span class="co">#&gt; alpha = 0.37: FDPhat 0.3684, Number of Rej. 38</span></span>
<span id="cb17-25"><a href="#cb17-25"></a><span class="co">#&gt; alpha = 0.36: FDPhat 0.3514, Number of Rej. 37</span></span>
<span id="cb17-26"><a href="#cb17-26"></a><span class="co">#&gt; alpha = 0.35: FDPhat 0.3125, Number of Rej. 16</span></span>
<span id="cb17-27"><a href="#cb17-27"></a><span class="co">#&gt; alpha = 0.34: FDPhat 0.3125, Number of Rej. 16</span></span>
<span id="cb17-28"><a href="#cb17-28"></a><span class="co">#&gt; alpha = 0.33: FDPhat 0.3125, Number of Rej. 16</span></span>
<span id="cb17-29"><a href="#cb17-29"></a><span class="co">#&gt; alpha = 0.32: FDPhat 0.3125, Number of Rej. 16</span></span>
<span id="cb17-30"><a href="#cb17-30"></a><span class="co">#&gt; alpha = 0.31: FDPhat 0.25, Number of Rej. 16</span></span>
<span id="cb17-31"><a href="#cb17-31"></a><span class="co">#&gt; alpha = 0.3: FDPhat 0.25, Number of Rej. 16</span></span>
<span id="cb17-32"><a href="#cb17-32"></a><span class="co">#&gt; alpha = 0.29: FDPhat 0.25, Number of Rej. 16</span></span>
<span id="cb17-33"><a href="#cb17-33"></a><span class="co">#&gt; alpha = 0.28: FDPhat 0.25, Number of Rej. 16</span></span>
<span id="cb17-34"><a href="#cb17-34"></a><span class="co">#&gt; alpha = 0.27: FDPhat 0.25, Number of Rej. 16</span></span>
<span id="cb17-35"><a href="#cb17-35"></a><span class="co">#&gt; alpha = 0.26: FDPhat 0.25, Number of Rej. 16</span></span>
<span id="cb17-36"><a href="#cb17-36"></a><span class="co">#&gt; alpha = 0.25: FDPhat 0.25, Number of Rej. 16</span></span>
<span id="cb17-37"><a href="#cb17-37"></a><span class="co">#&gt; alpha = 0.24: FDPhat 0.2, Number of Rej. 15</span></span>
<span id="cb17-38"><a href="#cb17-38"></a><span class="co">#&gt; alpha = 0.23: FDPhat 0.2, Number of Rej. 15</span></span>
<span id="cb17-39"><a href="#cb17-39"></a><span class="co">#&gt; alpha = 0.22: FDPhat 0.2, Number of Rej. 15</span></span>
<span id="cb17-40"><a href="#cb17-40"></a><span class="co">#&gt; alpha = 0.21: FDPhat 0.2, Number of Rej. 15</span></span>
<span id="cb17-41"><a href="#cb17-41"></a><span class="co">#&gt; alpha = 0.2: FDPhat 0.2, Number of Rej. 15</span></span>
<span id="cb17-42"><a href="#cb17-42"></a><span class="co">#&gt; alpha = 0.19: FDPhat 0.1429, Number of Rej. 14</span></span>
<span id="cb17-43"><a href="#cb17-43"></a><span class="co">#&gt; alpha = 0.18: FDPhat 0.1429, Number of Rej. 14</span></span>
<span id="cb17-44"><a href="#cb17-44"></a><span class="co">#&gt; alpha = 0.17: FDPhat 0.1429, Number of Rej. 14</span></span>
<span id="cb17-45"><a href="#cb17-45"></a><span class="co">#&gt; alpha = 0.16: FDPhat 0.1429, Number of Rej. 14</span></span>
<span id="cb17-46"><a href="#cb17-46"></a><span class="co">#&gt; alpha = 0.15: FDPhat 0.1429, Number of Rej. 14</span></span>
<span id="cb17-47"><a href="#cb17-47"></a><span class="co">#&gt; alpha = 0.14: FDPhat 0.0769, Number of Rej. 13</span></span>
<span id="cb17-48"><a href="#cb17-48"></a><span class="co">#&gt; alpha = 0.13: FDPhat 0.0769, Number of Rej. 13</span></span>
<span id="cb17-49"><a href="#cb17-49"></a><span class="co">#&gt; alpha = 0.12: FDPhat 0.0769, Number of Rej. 13</span></span>
<span id="cb17-50"><a href="#cb17-50"></a><span class="co">#&gt; alpha = 0.11: FDPhat 0.0769, Number of Rej. 13</span></span>
<span id="cb17-51"><a href="#cb17-51"></a><span class="co">#&gt; alpha = 0.1: FDPhat 0.0769, Number of Rej. 13</span></span>
<span id="cb17-52"><a href="#cb17-52"></a><span class="co">#&gt; alpha = 0.09: FDPhat 0.0769, Number of Rej. 13</span></span>
<span id="cb17-53"><a href="#cb17-53"></a><span class="co">#&gt; alpha = 0.08: FDPhat 0.0769, Number of Rej. 13</span></span></code></pre></div>
</div>
</div>
<div id="generic-interface-of-adaptmt-package" class="section level1">
<h1 class="hasAnchor">
<a href="#generic-interface-of-adaptmt-package" class="anchor"></a>Generic Interface of <code>adaptMT</code> Package</h1>
<p>The data-adaptive/human-in-the-loop part is embedded in the EM sub-pipeline:</p>
<embed src="flowchart_EM.pdf" width="450px" height="250px" style="display: block; margin: auto;" type="application/pdf"></embed><embed src="flowchart_EM_ms.pdf" width="500px" height="250px" style="display: block; margin: auto;" type="application/pdf"></embed><p>There are two main user-specified options: - exponential family to fit E-steps, characterized by input <code>dist</code>, of type <code>exp_family</code>; - functions to fit <span class="math inline">\(\pi_{1}(x)\)</span> and <span class="math inline">\(\mu(x)\)</span>, including those for initialization, characterized by input <code>models</code>, of type <code>adapt_model</code>.</p>
<p>We introduce each of them in the following subsections. For a complete list of other inputs and outputs, see <code><a href="../reference/adapt.html">?adapt</a></code>.</p>
<div id="exp_family-objects-for-fitting-e-steps" class="section level2">
<h2 class="hasAnchor">
<a href="#exp_family-objects-for-fitting-e-steps" class="anchor"></a><code>exp_family</code> objects for fitting E-steps</h2>
<p>An <code>exp_family</code> object is a list containing all required information for an exponential family, with mean-parameter <span class="math inline">\(\mu\)</span> and natural parameter <span class="math inline">\(\eta = \eta(\mu)\)</span> <span class="math display">\[h(p; \mu) = \exp\left\{(\eta(\mu) - \eta(\mu^{*}))g(p) - (A(\mu) - A(\mu^{*}))\right\}.\]</span> The above formulation exploits an over-parametrization by introducing <span class="math inline">\(\mu^{*}\)</span> to guarantee that <span class="math inline">\(U([0, 1])\)</span> belongs to the family and <span class="math inline">\(h(p; \mu^{*})\)</span> gives its density. Here by mean-parameter we mean by convention that <span class="math display">\[E_{\mu}[g(p)] = \mu.\]</span></p>
<p><code>adaptMT</code> package provides a generic function <code>exp_family</code> to construct an <code>exp_family</code> object from scratch. The required arguments include: (1) <code>g</code> as a function; (2) <code>ginv</code> as the inverse function of <code>g</code>; (3) <code>eta</code> as a function; (4) <code>mustar</code> as a numeric; (5) <code>A</code> as a function.</p>
<p>The current version of <code>adaptMT</code> package provides two instances: <code><a href="../reference/gen_exp_family.html">beta_family()</a></code> and <code><a href="../reference/gen_exp_family.html">inv_gaussian_family()</a></code>. For <code><a href="../reference/gen_exp_family.html">beta_family()</a></code>, <span class="math display">\[g(p) = -\log(p), \,\,\,\, g^{-1}(x) = e^{-x}, \,\,\,\, \eta(\mu) = -\frac{1}{\mu}, \,\,\,\, \mu^{*} = 1, \,\,\,\, A(\mu) = \log(\mu).\]</span> For <code><a href="../reference/gen_exp_family.html">inv_gaussian_family()</a></code>, <span class="math display">\[g(p) = \Phi^{-1}(1 - p), \,\,\,\, g^{-1}(x) = 1 - \Phi(x), \,\,\,\, \eta(\mu) = \mu, \,\,\,\, \mu^{*} = 0, \,\,\,\, A(\mu) = \frac{1}{2}\mu^{2}.\]</span> We found that <code><a href="../reference/gen_exp_family.html">beta_family()</a></code> consistently yields better results. Even in the case that the p-values are generated from a normal model as <code><a href="../reference/gen_exp_family.html">inv_gaussian_family()</a></code> specifies, <code><a href="../reference/gen_exp_family.html">beta_family()</a></code> still has comparable performance. For this reason, we take <code>dist = beta_family()</code> as the default input in <code>adapt</code> and all its wrappers.</p>
<p><code>adaptMT</code> package provides a generic function to generate an ‘exp_family’ object based on these inputs. For instance, the following code generates <code><a href="../reference/gen_exp_family.html">beta_family()</a></code>,</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="co"># g function</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>g &lt;-<span class="st"> </span><span class="cf">function</span>(x){</span>
<span id="cb18-3"><a href="#cb18-3"></a>    tmp &lt;-<span class="st"> </span><span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(x)</span>
<span id="cb18-4"><a href="#cb18-4"></a>    <span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">pmax</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">pmin</a></span>(tmp, <span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="dv">10</span><span class="op">^-</span><span class="dv">15</span>)), <span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="dv">1-10</span><span class="op">^-</span><span class="dv">15</span>))</span>
<span id="cb18-5"><a href="#cb18-5"></a>}</span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="co"># inverse function of g</span></span>
<span id="cb18-7"><a href="#cb18-7"></a>ginv &lt;-<span class="st"> </span><span class="cf">function</span>(x){</span>
<span id="cb18-8"><a href="#cb18-8"></a>    tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(<span class="op">-</span>x)</span>
<span id="cb18-9"><a href="#cb18-9"></a>    <span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">pmax</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">pmin</a></span>(tmp, <span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="dv">10</span><span class="op">^-</span><span class="dv">15</span>), <span class="dv">10</span><span class="op">^-</span><span class="dv">15</span>)</span>
<span id="cb18-10"><a href="#cb18-10"></a>}</span>
<span id="cb18-11"><a href="#cb18-11"></a><span class="co"># eta function (mapping from the mean parameter to the natural parameter)</span></span>
<span id="cb18-12"><a href="#cb18-12"></a>eta &lt;-<span class="st"> </span><span class="cf">function</span>(mu){<span class="op">-</span><span class="dv">1</span><span class="op">/</span>mu}</span>
<span id="cb18-13"><a href="#cb18-13"></a><span class="co"># mu* (the mean parameter corresponding to U([0, 1]))</span></span>
<span id="cb18-14"><a href="#cb18-14"></a>mustar &lt;-<span class="st"> </span><span class="dv">1</span></span>
<span id="cb18-15"><a href="#cb18-15"></a><span class="co"># A function (the partition function)</span></span>
<span id="cb18-16"><a href="#cb18-16"></a>A &lt;-<span class="st"> </span>log</span>
<span id="cb18-17"><a href="#cb18-17"></a><span class="co"># Set a name; optional</span></span>
<span id="cb18-18"><a href="#cb18-18"></a>name &lt;-<span class="st"> "beta"</span></span>
<span id="cb18-19"><a href="#cb18-19"></a><span class="co"># Set the default family as Gamma() (for glm, gam, glmnet, etc.); optional</span></span>
<span id="cb18-20"><a href="#cb18-20"></a>family &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/family.html">Gamma</a></span>()</span>
<span id="cb18-21"><a href="#cb18-21"></a><span class="co"># Generate beta_family()</span></span>
<span id="cb18-22"><a href="#cb18-22"></a>beta_family &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gen_exp_family.html">gen_exp_family</a></span>(g, ginv, eta, mustar, A, name, family)</span></code></pre></div>
</div>
<div id="adapt_model-objects-for-fitting-m-steps" class="section level2">
<h2 class="hasAnchor">
<a href="#adapt_model-objects-for-fitting-m-steps" class="anchor"></a><code>adapt_model</code> objects for fitting M-steps</h2>
<p>An <code>adapt_model</code> object is a list containing the fitting functions (and their arguments) in M-steps. It has three entries: <code>name</code>, <code>algo</code> and <code>args</code>. <code>name</code> gives the name of the model, which is optional; <code>algo</code> is a list of four functions <code>pifun</code>, <code>mufun</code>, <code>pifun_init</code> and <code>mufun_init</code>, for fitting <span class="math inline">\(\pi_{1}(x)\)</span> and <span class="math inline">\(\mu(x)\)</span> and initializing <span class="math inline">\(\pi_{1}(x)\)</span> and <span class="math inline">\(\mu(x)\)</span>, respectively; <code>args</code> is a list of four lists <code>piargs</code>, <code>muargs</code>, <code>piargs_init</code> and <code>muargs_init</code> that give the other arguments passed to <code>pifun</code>, <code>mufun</code>, <code>pifun_init</code> and <code>mufun_init</code>. <code>adaptMT</code> package provides a generic function that takes the aforementioned elements as input to generate an <code>adapt_model</code> object, that can be passed into <code>adapt</code> or its variants.</p>
<div id="design-principles-and-requirements-for-pifun-and-mufun" class="section level3">
<h3 class="hasAnchor">
<a href="#design-principles-and-requirements-for-pifun-and-mufun" class="anchor"></a>Design principles and requirements for <code>pifun</code> and <code>mufun</code>
</h3>
<p>Let <span class="math inline">\(P^{(r-1)}[\cdot \mid D_{t}]\)</span> denotes the posterior probability at step <span class="math inline">\(t\)</span> after <span class="math inline">\((r-1)\)</span> iteration in the EM algorithm. Here <span class="math inline">\(D_{t}\)</span> denotes the partially-masked dataset, namely <span class="math inline">\((x_i, \tilde{p}_{t, i})_{i\in [n]}\)</span>. Let <span class="math display">\[\hat{H}_{i}^{r} = P^{(r-1)} [H_{i} = 1\mid D_{t}], \quad \hat{b}_{i}^{r} = P^{(r-1)}[p_{i}\le 0.5 \mid D_{t}]\]</span> To fit <span class="math inline">\(\pi_{1}(x)\)</span> in the <span class="math inline">\(r\)</span>-th iteration, one can construct a pseudo-dataset with <span class="math display">\[y_{\pi}^{r} = (1, \ldots, 1, 0, \ldots, 0)\in R^{2n},\]</span> with weights <span class="math display">\[w_{\pi}^{r} = \left(\hat{H}_{1}^{r}, \ldots, \hat{H}_{n}^{r}, 1 - \hat{H}_{1}^{t}, \ldots, 1 - \hat{H}_{n}^{r}\right).\]</span> Let <span class="math inline">\(xx\)</span> be the covariate matrix after replicated for twice, namely <code><a href="https://rdrr.io/r/base/cbind.html">rbind(x, x)</a></code>. Then one can apply any classification algorithm on <span class="math inline">\((xx, y_{\pi}^{r})\)</span> with weights <span class="math inline">\(w_{\pi}^{r}\)</span> and estimate <span class="math inline">\(\hat{\pi}_{1}^{r}(x)\)</span> by the fitted probability of the first <span class="math inline">\(n\)</span> units.</p>
<p>Similarly, to fit <span class="math inline">\(\mu(x)\)</span> in the <span class="math inline">\(r\)</span>-th iteration, one can construct a pseudo-dataset with outcomes <span class="math display">\[y_{\mu}^{r} = (p_{1},\ldots, p_{n}, 1-p_{1}, \ldots, 1-p_{n}),\]</span> with weights <span class="math display">\[w_{\mu}^{r} = \left(\hat{b}_{1}^{r}\cdot \hat{H}_{1}^{r}, \ldots, \hat{b}_{n}^{r}\cdot \hat{H}_{n}^{r}, (1-\hat{b}_{1}^{r})\cdot \hat{H}_{1}^{r}, \ldots, (1-\hat{b}_{n}^{r})\cdot \hat{H}_{n}^{r}\right).\]</span> Then one can apply any algorithm on <span class="math inline">\((xx, y_{\mu}^{r})\)</span> with weights <span class="math inline">\(w_{\mu}^{r}\)</span> and estimate <span class="math inline">\(\hat{\mu}(x)\)</span> by the fitted values of the first <span class="math inline">\(n\)</span> units. Note that this is a more general version that Algorithm 2 in our paper. It is straightforward to show that for generalized linear models, this is equivalent to Algorithm 2. Also, as mentioned in Appendix A.3 in the supplementary material, we found that it yields a more stable performance by replacing the weights to <span class="math display">\[w_{\mu}^{r} = \left(\hat{b}_{1}^{r}, \ldots, \hat{b}_{n}^{r}, 1-\hat{b}_{1}^{r}, \ldots, 1-\hat{b}_{n}^{r}\right).\]</span></p>
<p>Note that <span class="math inline">\(\hat{H}_{i}^{r}\)</span> and <span class="math inline">\(\hat{b}_{i}^{r}\)</span> are generated within the algorithm so the user do not need to take care of them. Rather, the user only need to specifies the functions <code>pifun</code> and <code>mufun</code> as if <span class="math inline">\(\hat{H}_{i}^{r}\)</span> and <span class="math inline">\(\hat{b}_{i}^{r}\)</span> are known. More specifically, both <code>pifun</code> and <code>mufun</code> take <span class="math inline">\(x\)</span> and some output that has been calculated in the algorithm (<span class="math inline">\(y_{\pi}^{r}\)</span> or <span class="math inline">\(y_{\mu}^{r}\)</span>) with some weights that also has been calculated in the algorithm (<span class="math inline">\(w_{\pi}^{r}\)</span> or <span class="math inline">\(w_{\mu}^{r}\)</span>). They both output a sequence of estimates (<span class="math inline">\(\pi_{1}(x)\)</span> or <span class="math inline">\(\mu(x)\)</span>) of length <span class="math inline">\(n\)</span>, as well as some other information (e.g. degree of freedom, for performing model selection).</p>
<p>In <code>adaptMT</code> package, two types of inputs are accepted for <code>pifun</code> and <code>mufun</code>: <code><a href="https://rdrr.io/r/base/list.html">list(formula = , data = , weights = , ...)</a></code> or <code><a href="https://rdrr.io/r/base/list.html">list(x = , y = , weights = , ...)</a></code>. We refer to them as formula-data input type and x-y input type. Both of them should have ouput in the form of <code><a href="https://rdrr.io/r/base/list.html">list(fitv = , info = )</a></code> where <code>fitv</code> gives the fitted values and <code>info</code> is a list that gives other information, including degree-of-freedom (<code>info$df</code>), variable importance (<code>info$vi</code>), etc.. If model selection is performed, the entry <code>info$df</code> must be given.</p>
<div id="formula-data-input-format" class="section level4">
<h4 class="hasAnchor">
<a href="#formula-data-input-format" class="anchor"></a>formula-data input format</h4>
<p>This input type is adapted to a majority of fitting functions in <code>R</code> where the model is specified as a combination of a source data (<code>data</code>) and a formula (<code>formula</code>). In <code>pifun</code> and <code>mufun</code>, the user only needs to provide a partial formula without a response name. For instance, if the covariate <code>x</code> (in the input of <code>adapt</code>) is a data.frame that has a single column named “x” and one wants to fit <span class="math inline">\(\pi_{1}(x)\)</span> using a logistic-Gamma GLM with 8 knots, the formula should be specified as <code>"ns(x, df = 8)"</code> or <code>"~ ns(x, df = 8)"</code> or <code>~ ns (x, df = 8)</code>. Then <code>adapt</code> automatically construct a data.frame in the form of data.frame(x = xx, ** = …), where ** is a response name that is automatically generated by the algorithm which is guaranteed not to be conflicted with any variable names in <code>x</code>, and … is the outcome <span class="math inline">\(y_{\pi}^{r}\)</span> or <span class="math inline">\(y_{\mu}^{r}\)</span>. For technical reasons (unpleasant feature of <code>formula</code> objects in <code>R</code>), the input <code>formula</code> in <code>pifun</code> or <code>mufun</code> is always <strong>assumed to be a character</strong>. However, this technical issue can be simply avoid by adding a line at the beginning of <code>pifun</code> or <code>mufun</code>. See the example below.</p>
<p>For instance, in logistic-Gamma GLM, <code>pifun</code> is a logistic regression (of <span class="math inline">\(y_{\pi}^{r}\)</span> on <span class="math inline">\(xx\)</span> with weights <span class="math inline">\(w_{\pi}^{r}\)</span>) and <code>mufun</code> is a Gamma GLM (or <span class="math inline">\(y_{\mu}^{r}\)</span> on <span class="math inline">\(xx\)</span> with weights <span class="math inline">\(w_{\mu}^{r}\)</span>). The exemplary code is given as follows:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>pifun_glm &lt;-<span class="st"> </span><span class="cf">function</span>(formula, data, weights, ...){</span>
<span id="cb19-2"><a href="#cb19-2"></a>  formula &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span>(formula) <span class="co"># Transform the argument `formula` from character to formula</span></span>
<span id="cb19-3"><a href="#cb19-3"></a>  fit &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span>(formula, data, <span class="dt">weights =</span> weights,</span>
<span id="cb19-4"><a href="#cb19-4"></a>                    <span class="dt">family =</span> <span class="kw"><a href="https://rdrr.io/r/stats/family.html">quasibinomial</a></span>(), ...) </span>
<span id="cb19-5"><a href="#cb19-5"></a>  <span class="co"># Here quasibinomial() family gets rid of meaningless warning messages</span></span>
<span id="cb19-6"><a href="#cb19-6"></a>  fitv &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(fit, <span class="dt">type =</span> <span class="st">"response"</span>) <span class="co"># fitted values for pi1(x)</span></span>
<span id="cb19-7"><a href="#cb19-7"></a>  info &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">df =</span> fit<span class="op">$</span>rank) <span class="co"># degree-of-freedom being the rank of the design matrix</span></span>
<span id="cb19-8"><a href="#cb19-8"></a>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">fitv =</span> fitv, <span class="dt">info =</span> info))</span>
<span id="cb19-9"><a href="#cb19-9"></a>}</span>
<span id="cb19-10"><a href="#cb19-10"></a></span>
<span id="cb19-11"><a href="#cb19-11"></a>mufun_glm &lt;-<span class="st"> </span><span class="cf">function</span>(formula, data, weights, ...){</span>
<span id="cb19-12"><a href="#cb19-12"></a>  formula &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span>(formula) <span class="co"># Transform the argument `formula` from character to formula</span></span>
<span id="cb19-13"><a href="#cb19-13"></a>  fit &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span>(formula, data, <span class="dt">weights =</span> weights,</span>
<span id="cb19-14"><a href="#cb19-14"></a>                <span class="dt">family =</span> <span class="kw"><a href="https://rdrr.io/r/stats/family.html">Gamma</a></span>(), ...)</span>
<span id="cb19-15"><a href="#cb19-15"></a>  fitv &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(fit, <span class="dt">type =</span> <span class="st">"response"</span>) <span class="co"># fitted values for mu(x)</span></span>
<span id="cb19-16"><a href="#cb19-16"></a>  info &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">df =</span> fit<span class="op">$</span>rank) <span class="co"># degree-of-freedom being the rank of the design matrix</span></span>
<span id="cb19-17"><a href="#cb19-17"></a>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">fitv =</span> fitv, <span class="dt">info =</span> info))</span>
<span id="cb19-18"><a href="#cb19-18"></a>}</span></code></pre></div>
<p>Note that the above code is a over-simplified version of the implementation in <code>adaptMT</code> package because of the subtlety in fitting GLM using <code>glm</code> function in R <code>stats</code> package with <code>inverse</code> or <code>log</code> link functions (see <code><a href="https://rdrr.io/r/stats/glm.html">?glm</a></code> and <code><a href="https://rdrr.io/r/stats/family.html">?family</a></code> for details), for which the initialization might fail.</p>
</div>
<div id="x-y-input-format" class="section level4">
<h4 class="hasAnchor">
<a href="#x-y-input-format" class="anchor"></a>x-y input format</h4>
<p>This input type is adapted to a fraction of methods, especially for modern machine learning algorithms, where the model is specified by the outcome variable (<code>y</code>) and a covariate matrix (<code>x</code>). For instance, <code>glmnet</code> package falls in this category, though the <code>glmnetUtils</code> package provides a formula interface for it.</p>
<p>To illustrate, we consider the <span class="math inline">\(L_1\)</span>-penalized logistic-Gamma GLM by fitting the same aforementioned regressions for <span class="math inline">\(\pi_{1}(x)\)</span> and <span class="math inline">\(\mu(x)\)</span> but with a <span class="math inline">\(L_1\)</span> penalty and with the penalty level selected by cross-validation. Note that <code>glmnet</code> package does not support Gamma GLM, we use <code>HDtweedie</code> package instead; See <code>?HDtweedie</code> for details. The exemplary code is given as follows:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>pifun_glmnet &lt;-<span class="st"> </span><span class="cf">function</span>(x, y, weights, ...){</span>
<span id="cb20-2"><a href="#cb20-2"></a>  fit &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/glmnet/man/cv.glmnet.html">cv.glmnet</a></span>(x, y, <span class="dt">weights =</span> weights, <span class="dt">family =</span> <span class="st">"binomial"</span>, ...) </span>
<span id="cb20-3"><a href="#cb20-3"></a>  fitv &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(fit, <span class="dt">newx =</span> x, <span class="dt">s =</span> <span class="st">"lambda.min"</span>, <span class="dt">type =</span> <span class="st">"response"</span>) <span class="co"># fitted values for pi1(x)</span></span>
<span id="cb20-4"><a href="#cb20-4"></a>  beta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span>(fit, <span class="dt">s =</span> <span class="st">"lambda.min"</span>)</span>
<span id="cb20-5"><a href="#cb20-5"></a>  info &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">df =</span> <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(beta <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>)) <span class="co"># degree-of-freedom being the number of non-zero coefficients</span></span>
<span id="cb20-6"><a href="#cb20-6"></a>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">fitv =</span> fitv, <span class="dt">info =</span> info))</span>
<span id="cb20-7"><a href="#cb20-7"></a>}</span>
<span id="cb20-8"><a href="#cb20-8"></a></span>
<span id="cb20-9"><a href="#cb20-9"></a>mufun_glmnet &lt;-<span class="st"> </span><span class="cf">function</span>(x, y, weights, ...){</span>
<span id="cb20-10"><a href="#cb20-10"></a>  fit &lt;-<span class="st"> </span>HDtweedie<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/HDtweedie/man/cv.HDtweedie.html">cv.HDtweedie</a></span>(x, y, <span class="dt">p =</span> <span class="dv">2</span>, <span class="dt">weights =</span> weights, ...)</span>
<span id="cb20-11"><a href="#cb20-11"></a>  fitv &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(fit, <span class="dt">newx =</span> x, <span class="dt">s =</span> <span class="st">"lambda.min"</span>, <span class="dt">type =</span> <span class="st">"response"</span>) <span class="co"># fitted values for mu(x)</span></span>
<span id="cb20-12"><a href="#cb20-12"></a>  beta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span>(fit, <span class="dt">s =</span> <span class="st">"lambda.min"</span>)</span>
<span id="cb20-13"><a href="#cb20-13"></a>  info &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">df =</span> <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(beta <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>)) <span class="co"># degree-of-freedom being the number of non-zero coefficients</span></span>
<span id="cb20-14"><a href="#cb20-14"></a>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">fitv =</span> fitv, <span class="dt">info =</span> info))</span>
<span id="cb20-15"><a href="#cb20-15"></a>}</span></code></pre></div>
<p>Again the above code is a over-simplified for the purpose of illustration. We deal with the initialization and other numerical issues much more carefully in the real implementation.</p>
</div>
</div>
<div id="design-principles-and-requirements-for-pifun_init-and-mufun_init" class="section level3">
<h3 class="hasAnchor">
<a href="#design-principles-and-requirements-for-pifun_init-and-mufun_init" class="anchor"></a>Design principles and requirements for <code>pifun_init</code> and <code>mufun_init</code>
</h3>
<p><code>pifun_init</code> and <code>mufun_init</code> give the functions to initialize <span class="math inline">\(\pi_{1}(x)\)</span> and <span class="math inline">\(\mu(x)\)</span>, respectively. Both of them are required to take at least three inputs: <code>x</code>, <code>pvals</code> and <code>s</code>. They are required output <code><a href="https://rdrr.io/r/base/list.html">list(fitv = , info = )</a></code>, with the same meaning as in <code>pifun</code> and <code>mufun</code>.</p>
<p>As described in <span class="citation">(Lei and Fithian 2016)</span>, the initialization of <span class="math inline">\(\pi_{1}(x)\)</span> is done by defining <span class="math display">\[J_{i} = 1 - \frac{I(s_{0}(x_{i}) \le p_{i}\le 1 - s_{0}(x_{i}))}{1 - 2s_{0}(x_i)}\]</span> and regress <span class="math inline">\(J_{i}\)</span> on <span class="math inline">\(x_{i}\)</span> with the same formula as in <code>pifun</code>. For <code>mufun_init</code> we simply impute the masked pvalues ad-hocly by <span class="math inline">\(\min\{p_i, 1 - p_i\}\)</span> and run <code>mufun</code> on the imputed data. The following chunk gives the exemplary code for <code>adapt_glm</code>. Note that the code is a over-simplified version of the implementation in <code>adaptMT</code> package.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>pifun_init_glm &lt;-<span class="st"> </span><span class="cf">function</span>(x, pvals, s, </span>
<span id="cb21-2"><a href="#cb21-2"></a>                       formula, ...){</span>
<span id="cb21-3"><a href="#cb21-3"></a>    <span class="co"># The input `formula` is a half-formula of type character, e.g. ~ ns(x, df = 8)</span></span>
<span id="cb21-4"><a href="#cb21-4"></a>    J &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(</span>
<span id="cb21-5"><a href="#cb21-5"></a>        pvals <span class="op">&lt;</span><span class="st"> </span>s <span class="op">|</span><span class="st"> </span>pvals <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>s, <span class="dv">1</span>,</span>
<span id="cb21-6"><a href="#cb21-6"></a>        <span class="dv">2</span> <span class="op">*</span><span class="st"> </span>s <span class="op">/</span><span class="st"> </span>(<span class="dv">2</span> <span class="op">*</span><span class="st"> </span>s <span class="op">-</span><span class="st"> </span><span class="dv">1</span>)</span>
<span id="cb21-7"><a href="#cb21-7"></a>        )</span>
<span id="cb21-8"><a href="#cb21-8"></a></span>
<span id="cb21-9"><a href="#cb21-9"></a>    <span class="co"># Complete the formula by adding the response variable J</span></span>
<span id="cb21-10"><a href="#cb21-10"></a>    complete_formula &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"J"</span>, formula)) </span>
<span id="cb21-11"><a href="#cb21-11"></a>    data &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(x, <span class="dt">J =</span> J)</span>
<span id="cb21-12"><a href="#cb21-12"></a>    mod &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span>(complete_formula, <span class="dt">family =</span> <span class="kw"><a href="https://rdrr.io/r/stats/family.html">gaussian</a></span>(), <span class="dt">data =</span> data)</span>
<span id="cb21-13"><a href="#cb21-13"></a>    </span>
<span id="cb21-14"><a href="#cb21-14"></a>    fitv &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(mod, <span class="dt">type =</span> <span class="st">"response"</span>))</span>
<span id="cb21-15"><a href="#cb21-15"></a>    info &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">df =</span> fit<span class="op">$</span>rank)</span>
<span id="cb21-16"><a href="#cb21-16"></a>    </span>
<span id="cb21-17"><a href="#cb21-17"></a>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">fitv =</span> fitv, <span class="dt">info =</span> info))</span>
<span id="cb21-18"><a href="#cb21-18"></a>}</span>
<span id="cb21-19"><a href="#cb21-19"></a></span>
<span id="cb21-20"><a href="#cb21-20"></a>mufun_init_glm &lt;-<span class="st"> </span><span class="cf">function</span>(x, pvals, s, ...){</span>
<span id="cb21-21"><a href="#cb21-21"></a>    phat &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(</span>
<span id="cb21-22"><a href="#cb21-22"></a>        pvals <span class="op">&lt;</span><span class="st"> </span>s <span class="op">|</span><span class="st"> </span>pvals <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>s,</span>
<span id="cb21-23"><a href="#cb21-23"></a>        <span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">pmin</a></span>(pvals, <span class="dv">1</span> <span class="op">-</span><span class="st"> </span>pvals),</span>
<span id="cb21-24"><a href="#cb21-24"></a>        pvals</span>
<span id="cb21-25"><a href="#cb21-25"></a>        )</span>
<span id="cb21-26"><a href="#cb21-26"></a>    yhat &lt;-<span class="st"> </span>dist<span class="op">$</span><span class="kw">g</span>(phat)</span>
<span id="cb21-27"><a href="#cb21-27"></a></span>
<span id="cb21-28"><a href="#cb21-28"></a>    <span class="co"># Complete the formula by adding the response variable yhat</span></span>
<span id="cb21-29"><a href="#cb21-29"></a>    complete_formula &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"yhat"</span>, formula)) </span>
<span id="cb21-30"><a href="#cb21-30"></a>    data &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(x, <span class="dt">yhat =</span> yhat)</span>
<span id="cb21-31"><a href="#cb21-31"></a>    mod &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span>(complete_formula, <span class="dt">family =</span> <span class="kw"><a href="https://rdrr.io/r/stats/family.html">Gamma</a></span>(), <span class="dt">data =</span> data)</span>
<span id="cb21-32"><a href="#cb21-32"></a>    </span>
<span id="cb21-33"><a href="#cb21-33"></a>    fitv &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(mod, <span class="dt">type =</span> <span class="st">"response"</span>))</span>
<span id="cb21-34"><a href="#cb21-34"></a>    info &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">df =</span> fit<span class="op">$</span>rank)</span>
<span id="cb21-35"><a href="#cb21-35"></a>    </span>
<span id="cb21-36"><a href="#cb21-36"></a>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">fitv =</span> fitv, <span class="dt">info =</span> info))</span>
<span id="cb21-37"><a href="#cb21-37"></a>}</span></code></pre></div>
</div>
<div id="constructing-adapt_model-objects-by-gen_adapt_model" class="section level3">
<h3 class="hasAnchor">
<a href="#constructing-adapt_model-objects-by-gen_adapt_model" class="anchor"></a>Constructing <code>adapt_model</code> objects by <code>gen_adapt_model</code>
</h3>
<p><code>adaptMT</code> package provides a generic function <code>gen_adapt_model</code> to construct <code>adapt_model</code> objects. The code to generate an <code>adapt_model</code> object characterizing a Logistic-Gamma GLM on the spline bases of a univariate covariate <span class="math inline">\(x\)</span> with <span class="math inline">\(5\)</span> equi-quantile knots to fit <span class="math inline">\(\pi_{1}(x)\)</span> and <span class="math inline">\(\mu(x)\)</span>.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"splines"</span>) </span>
<span id="cb22-2"><a href="#cb22-2"></a>piargs &lt;-<span class="st"> </span>muargs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">formula =</span> <span class="st">"~ ns(x, df = 5)"</span>)</span>
<span id="cb22-3"><a href="#cb22-3"></a>adapt_model_glm_example &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gen_adapt_model.html">gen_adapt_model</a></span>(</span>
<span id="cb22-4"><a href="#cb22-4"></a>  <span class="dt">pifun =</span> pifun_glm, <span class="dt">mufun =</span> mufun_glm, </span>
<span id="cb22-5"><a href="#cb22-5"></a>  <span class="dt">pifun_init =</span> pifun_init_glm, <span class="dt">mufun_init =</span> mufun_init_glm,</span>
<span id="cb22-6"><a href="#cb22-6"></a>  <span class="dt">piargs =</span> piargs, <span class="dt">muargs =</span> muargs,</span>
<span id="cb22-7"><a href="#cb22-7"></a>  <span class="dt">piargs_init =</span> piargs, <span class="dt">muargs_init =</span> muargs</span>
<span id="cb22-8"><a href="#cb22-8"></a>)</span></code></pre></div>
<p><code>adaptMT</code> package also provides convenient ways to generate <code>adapt_models</code> for GLMs, GAMs and glmnets by just specifying <code>piargs</code> and <code>muargs</code>.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>adapt_model_glm_example2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gen_adapt_model.html">gen_adapt_model</a></span>(</span>
<span id="cb23-2"><a href="#cb23-2"></a>  <span class="dt">name =</span> <span class="st">"glm"</span>,</span>
<span id="cb23-3"><a href="#cb23-3"></a>  <span class="dt">piargs =</span> piargs, <span class="dt">muargs =</span> muargs</span>
<span id="cb23-4"><a href="#cb23-4"></a>)</span></code></pre></div>
<p>This yields the same <code>adapt_model</code> object that is implicitly used in <code>adapt_glm</code>. We show the equivalence on a toy dataset below.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(estrogen)</span>
<span id="cb24-2"><a href="#cb24-2"></a>pvals &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(estrogen<span class="op">$</span>pvals)</span>
<span id="cb24-3"><a href="#cb24-3"></a>x &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(estrogen<span class="op">$</span>ord_high))</span>
<span id="cb24-4"><a href="#cb24-4"></a>dist &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gen_exp_family.html">beta_family</a></span>()</span>
<span id="cb24-5"><a href="#cb24-5"></a>inds &lt;-<span class="st"> </span>(x<span class="op">$</span>x <span class="op">&lt;=</span><span class="st"> </span><span class="dv">1000</span>)</span>
<span id="cb24-6"><a href="#cb24-6"></a>pvals &lt;-<span class="st"> </span>pvals[inds]</span>
<span id="cb24-7"><a href="#cb24-7"></a>x &lt;-<span class="st"> </span>x[inds,,drop =<span class="st"> </span><span class="ot">FALSE</span>]</span>
<span id="cb24-8"><a href="#cb24-8"></a></span>
<span id="cb24-9"><a href="#cb24-9"></a>res1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/adapt_glm.html">adapt_glm</a></span>(<span class="dt">x =</span> x, <span class="dt">pvals =</span> pvals, <span class="dt">pi_formulas =</span> <span class="st">"~ ns(x, df = 5)"</span>, <span class="dt">mu_formulas =</span> <span class="st">"~ ns(x, df = 5)"</span>)</span>
<span id="cb24-10"><a href="#cb24-10"></a><span class="co">#&gt; alpha = 0.09: FDPhat 0.0891, Number of Rej. 842</span></span>
<span id="cb24-11"><a href="#cb24-11"></a><span class="co">#&gt; alpha = 0.08: FDPhat 0.0791, Number of Rej. 834</span></span>
<span id="cb24-12"><a href="#cb24-12"></a><span class="co">#&gt; alpha = 0.07: FDPhat 0.0699, Number of Rej. 815</span></span>
<span id="cb24-13"><a href="#cb24-13"></a><span class="co">#&gt; alpha = 0.06: FDPhat 0.059, Number of Rej. 797</span></span>
<span id="cb24-14"><a href="#cb24-14"></a><span class="co">#&gt; alpha = 0.05: FDPhat 0.0499, Number of Rej. 741</span></span>
<span id="cb24-15"><a href="#cb24-15"></a><span class="co">#&gt; alpha = 0.04: FDPhat 0.0397, Number of Rej. 705</span></span>
<span id="cb24-16"><a href="#cb24-16"></a><span class="co">#&gt; alpha = 0.03: FDPhat 0.0299, Number of Rej. 669</span></span>
<span id="cb24-17"><a href="#cb24-17"></a><span class="co">#&gt; alpha = 0.02: FDPhat 0.0193, Number of Rej. 569</span></span>
<span id="cb24-18"><a href="#cb24-18"></a>res2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/adapt.html">adapt</a></span>(<span class="dt">x =</span> x, <span class="dt">pvals =</span> pvals, <span class="dt">models =</span> adapt_model_glm_example2)</span>
<span id="cb24-19"><a href="#cb24-19"></a><span class="co">#&gt; alpha = 0.09: FDPhat 0.0891, Number of Rej. 842</span></span>
<span id="cb24-20"><a href="#cb24-20"></a><span class="co">#&gt; alpha = 0.08: FDPhat 0.0791, Number of Rej. 834</span></span>
<span id="cb24-21"><a href="#cb24-21"></a><span class="co">#&gt; alpha = 0.07: FDPhat 0.0699, Number of Rej. 815</span></span>
<span id="cb24-22"><a href="#cb24-22"></a><span class="co">#&gt; alpha = 0.06: FDPhat 0.059, Number of Rej. 797</span></span>
<span id="cb24-23"><a href="#cb24-23"></a><span class="co">#&gt; alpha = 0.05: FDPhat 0.0499, Number of Rej. 741</span></span>
<span id="cb24-24"><a href="#cb24-24"></a><span class="co">#&gt; alpha = 0.04: FDPhat 0.0397, Number of Rej. 705</span></span>
<span id="cb24-25"><a href="#cb24-25"></a><span class="co">#&gt; alpha = 0.03: FDPhat 0.0299, Number of Rej. 669</span></span>
<span id="cb24-26"><a href="#cb24-26"></a><span class="co">#&gt; alpha = 0.02: FDPhat 0.0193, Number of Rej. 569</span></span>
<span id="cb24-27"><a href="#cb24-27"></a><span class="kw"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(res1<span class="op">$</span>s, res2<span class="op">$</span>s)</span>
<span id="cb24-28"><a href="#cb24-28"></a><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
</div>
</div>
<div id="bioliography" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#bioliography" class="anchor"></a>Bioliography</h1>
<div id="refs" class="references hanging-indent">
<div id="ref-glmnet">
<p>Hastie, Trevor, and Junyang Qian. 2014. “Glmnet Vignette.” Technical report, Stanford.</p>
</div>
<div id="ref-lei2016adapt">
<p>Lei, Lihua, and William Fithian. 2016. “AdaPT: An Interactive Procedure for Multiple Testing with Side Information.” <em>arXiv Preprint arXiv:1609.06035</em>.</p>
</div>
<div id="ref-GAM">
<p>Wood, Simon N. 2006. <em>Generalized Additive Models: An Introduction with R</em>. Chapman; Hall/CRC.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#convenient-wrappers-in-adaptmt-package">Convenient Wrappers in <code>adaptMT</code> Package</a>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#using-adapt_glm">Using <code>adapt_glm</code></a></li>
      <li><a href="#using-adapt_gam">Using <code>adapt_gam</code></a></li>
      <li><a href="#using-adapt_glmnet">Using <code>adapt_glmnet</code></a></li>
      </ul>
</li>
      <li>
<a href="#generic-interface-of-adaptmt-package">Generic Interface of <code>adaptMT</code> Package</a>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#exp_family-objects-for-fitting-e-steps"><code>exp_family</code> objects for fitting E-steps</a></li>
      <li><a href="#adapt_model-objects-for-fitting-m-steps"><code>adapt_model</code> objects for fitting M-steps</a></li>
      </ul>
</li>
      <li><a href="#bioliography">Bioliography</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Lihua Lei.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
